src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs

```diff
@@ -244,14 +262,103 @@ public void OnEventCommand(EventCommandEventArgs command)
                         return;
                     }
 #endif
-                    if (command.Command == EventCommand.Update || command.Command == EventCommand.Disable ||
-                        command.Command == EventCommand.Enable)
+
+                    string commandSessionId = GetSessionId(command);
+
+                    if ((command.Command == EventCommand.Update
+                        || command.Command == EventCommand.Disable
+                        || command.Command == EventCommand.Enable)
+                        && _aggregationManager != null)
                     {
-                        if (_aggregationManager != null)
+                        if (command.Command == EventCommand.Update
+                            || command.Command == EventCommand.Enable)
+                        {
+                            IncrementRefCount(commandSessionId, command);
```

Comment:
@davmason - EventSource sends Enable events even when sessions are being disabled right, except for the last one? That makes it impossible to do this kind of ref-counting and why we had to resort to testing IsEnabled() instead for EventCounters as I recall.

Is behavior documented anywhere? I always struggle to remember it, nobody guesses it because its unintuitive and pages where I'd expect to find it don't mention it:
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventcommand?view=net-7.0
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource.oneventcommand?view=net-7.0#system-diagnostics-tracing-eventsource-oneventcommand(system-diagnostics-tracing-eventcommandeventargs)

Reply:
I'm leaving this as unresolved for now (I'm not sure if @davmason had a chance to take a look), but assuming the `ClientId` protocol is followed by all consumers, is there a reason why this can't/shouldn't work? In my own testing the ref counting seems to work properly (I tested with DM metrics, DM live metrics, and Dotnet Counters simultaneously), but I may be lacking some depth in understanding around this issue.

Reply:
Sorry, I didn't see Noah's comment. The underlying providers (ETW and EventPipe) send an enable command for every disabled session except the one that turns the provider off completely, but EventPipe will translate those to Enables so you don't have to worry about it.

The problem with ref counting is that we can't guarantee an even number of Enables and Disables with EventListeners. An EventListener can send as many Enables as they want and will send at most one Disable.

Reply:
For EventCounters we just check on every Disable command if the EventSource is active and shut everything down if it is not